exceptions/exceptions.hh

```diff
@@ -71,6 +71,7 @@ enum class exception_code : int32_t {
     // Scylla-specific codes
     // Allocated backwards from 0x1aff-0x1a01 to minimize the chance of collision with Cassandra.
     OVERFLOW_ERROR  = 0x1aff,
+    CDC_PREIMAGE_READ_FAILURE = 0x1aef,
```

Comment:
Extending a binary protocol like that is a slippery slope. I see that we alraedy did it even before that patch, but what is it good for if we do not write our own drivers? How do we know that all the drivers will handle it gracefully? At the very least we should not use them unless we negotiated that our own driver is in use.

Reply:
Interestingly we need a similar extension for LWT, Gleb :)

Reply:
(and I agree that extending the protocol on the server side only is useless)

Reply:
We do have an ability to negotiate features with our drivers. It is crazy how we are adding those two codes here without any regards with ABI compatibility.  Even Cassandra when they introduce new error codes only use them when a proper binary version of a protocol is negotiated. How are _we_ going to know which of our version supports what errors?

Reply:
@gleb-cloudius @kostja So, maybe I'll return `READ_FAILURE` or `UNAVAILABLE`?
Not that reusing codes can also lead to a mess (I mean, a less visible mess but potentially far-reaching).

Anyway, please, let's state a definitive opinion here. This is a simple patch and doesn't deserve to hang for that long.

Reply:
Please return READ_FALIURE and let's get this patch merged.

Reply:
You can report what exactly failed in the message supplied to the error.

Reply:
> Please return READ_FALIURE and let's get this patch merged.

Ooops, I cannot. Both `READ_FAILURE` and `UNAVAILABLE` impose constraints on the error message. Just checked: `request_execution_exception` with any of these crashes the python client.

My bad, the problem is still open.

Reply:
What do you mean? How returning READ_FAILURE crashes python driver?

Reply:
```
Traceback (most recent call last):
  File "/home/julex/scylla-tools-java/bin/cqlsh.py", line 1053, in perform_simple_statement
    result = future.result()
  File "/home/julex/scylla-tools-java/bin/../lib/cassandra-driver-internal-only-3.11.0-bb96859b.zip/cassandra-driver-3.11.0-bb96859b/cassandra/cluster.py", line 3925, in result
    raise self._final_exception
error: unpack requires a string argument of length 2
```
Same thing with `UNAVAILABLE`. When I throw plain `cassandra_exception` with those codes effect is identical. I didn't investigate what exactly happens on clientside, but it seems that these codes imply some structure in error response.

Reply:
Hm ...

If clients will not handle the error - it will be an issue (since error handling will not work as well - e.g. retries)

So looking at https://github.com/apache/cassandra/blob/trunk/doc/native_protocol_v5.spec there are some additional extensions for CDC. 

We do not support v5 protocol - so once we do support that we will have the ability to return CDC specific codes (nothing matches the preimage unfortunately as Cassandra does not do pre-image) 

So lets settle on something that does work.

Reply:
@jul-stas I've seen this before. It meas that the error encoding did not match error code. Make sure that you do not lose exception type between the place it was thrown and cqls server where it is encoded into a protocol message.